src/lib/ilqp.js

```diff
@@ -0,0 +1,180 @@
+
+      if (data.method === method) {
+        debug('response of type', method)
+        plugin.removeListener('incoming_message', onIncomingMessage)
+        resolve(response)
+      }
+    }
+
+    // TODO: optimize to not add a listener each time?
+    plugin.on('incoming_message', onIncomingMessage)
+  })
+
+  return Promise.race([
+    plugin.sendMessage(message).then(() => responded),
+    wait(timeout || 5000)
+      .then(() => { throw new Error('quote request timed out') })
+  ])
+}
+
+function _getQuote ({
+  plugin,
+  connector,
+  quoteQuery,
+  timeout
+}) {
+  const prefix = plugin.getInfo().prefix
+
+  debug('remote quote connector=' + connector, 'query=' + JSON.stringify(quoteQuery))
+  return _sendAndReceiveMessage({
+    plugin: plugin,
+    method: 'quote_response',
+    timeout: timeout,
+    message: {
+      ledger: prefix,
+      account: connector,
+      data: {
+        method: 'quote_request',
+        data: quoteQuery
+      }
+    }
+  }).then((response) => {
+    return response.data.data
+  }).catch((err) => {
+    // TODO: should this do something else about errors?
+    debug('ignoring remote quote error:', err.message)
+  })
+}
+
+function _getCheaperQuote (quote1, quote2) {
+  const source1 = new BigNumber(quote1.source_amount)
+  const dest1 = new BigNumber(quote1.destination_amount)
+
+  if (source1.lessThan(quote2.source_amount)) return quote1
+  if (dest1.greaterThan(quote2.destination_amount)) return quote1
+
+  return quote2
+}
+
+/**
+  * @param {Object} plugin The LedgerPlugin used to send quote request
+  * @param {Object} query
+  * @param {String} query.sourceAddress Sender's address
+  * @param {String} query.destinationAddress Recipient's address
+  * @param {String} [query.sourceAmount] Either the sourceAmount or destinationAmount must be specified
+  * @param {String} [query.destinationAmount] Either the sourceAmount or destinationAmount must be specified
+  * @param {String|Number} [query.sourceExpiryDuration] Number of seconds between when the source transfer is proposed and when it expires.
+  * @param {String|Number} [query.destinationExpiryDuration] Number of seconds between when the destination transfer is proposed and when it expires.
+  * @param {Object} [query.destinationPrecision] Precision of destination ledger
+  * @param {Object} [query.destinationScale] Scale of destination ledger
+  * @param {Array} [query.connectors] List of ILP addresses of connectors to use for this quote.
+  * @returns {Promise<Quote>}
+  */
+function * quote (plugin, {
+  sourceAddress,
+  destinationAddress,
+  sourceAmount,
+  destinationAmount,
+  sourceExpiryDuration,
+  destinationExpiryDuration,
+  destinationPrecision,
+  destinationScale,
+  connectors,
+  timeout
+}) {
+  if (!xor(sourceAmount, destinationAmount)) {
+    throw new Error('should provide source or destination amount but not both' +
+      ' ' + JSON.stringify({ sourceAmount, destinationAmount }))
+  }
+
+  yield plugin.connect()
```

Comment:
maybe connection timeout ought to be a constructor property rather than one of the connect function, so that the user can set it without having to add a field to every function that calls connect